pkg/master/manager.go

```diff
@@ -0,0 +1,167 @@
+	"github.com/go-ping/ping"
+	"github.com/medik8s/self-node-remediation/pkg/peers"
+	corev1 "k8s.io/api/core/v1"
+	"os/exec"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
+)
+
+const (
+	initErrorText                = "error initializing master handler"
+	kubeletPort                  = "10250"
+	endpointToAccess             = "www.google.com"
+)
+
+var (
+	initError    = errors.New(initErrorText)
+	processError = errors.New("an error occurred during master remediation process")
+)
+
+//Manager contains logic and info needed to fence and remediate master nodes
+type Manager struct {
+	nodeName                             string
+	nodeRole                             peers.Role
+	wasEndpointAccessibleAtStart         bool
+	client                               client.Client
+	log                                  logr.Logger
+}
+
+//NewManager inits a new Manager return nil if init fails
+func NewManager(nodeName string, myClient client.Client) *Manager {
+	return &Manager{
+		nodeName:                     nodeName,
+		client:                       myClient,
+		wasEndpointAccessibleAtStart: false,
+		log:                          ctrl.Log.WithName("master").WithName("Manager"),
+	}
+}
+
+func (manager *Manager) Start(ctx context.Context) error {
+	if err := manager.initializeManager(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (manager *Manager) IsMaster() bool {
+	return manager.nodeRole == peers.Master
+}
+
+func (manager *Manager) IsMasterHealthy(workerPeerResponse peers.Response, isOtherMastersCanBeReached bool) bool {
+	switch workerPeerResponse.Reason {
+	//reported unhealthy by worker peers
+	case peers.UnHealthyBecauseCRFound:
+		return false
+	case peers.UnHealthyBecauseNodeIsIsolated:
+		return isOtherMastersCanBeReached
+	//reported healthy by worker peers
+	case peers.HealthyBecauseErrorsThresholdNotReached, peers.HealthyBecauseCRNotFound:
+		return true
+	//master node has connection to most workers, we assume it's not isolated (or at least that the master node that does not have worker peers quorum will reboot)
+	case peers.HealthyBecauseMostPeersCantAccessAPIServer:
+		return manager.isDiagnosticsPassed()
+	case peers.HealthyBecauseNoPeersWereFound:
+		return manager.isDiagnosticsPassed() && isOtherMastersCanBeReached
+
+	default:
+		manager.log.Error(processError, "node is considered unhealthy by worker peers for an unknown reason", "reason", workerPeerResponse.Reason, "node name", manager.nodeName)
+		return false
+	}
+
+}
+
+func (manager *Manager) isDiagnosticsPassed() bool {
+	if manager.isEndpointAccessLost() {
+		return false
+	} else if !manager.isKubeletServiceRunning() {
+		return false
+	} else if !manager.isEtcdRunning() {
+		return false
+	}
+	return true
+}
+
+func wrapWithInitError(err error) error {
+	return fmt.Errorf(initErrorText+" [%w]", err)
+}
+
+func (manager *Manager) initializeManager() error {
+	nodesList := &corev1.NodeList{}
```

Comment:
I'm already using The node name (fetched from MY_NODE_NAME) in order to pick the correct node from the list.
IIUC fetching our node will require having the `namespace` as well (opposed to fetching all the Nodes).

It's not a big deal to get the namespace (probably just fetch it from the env var or pass it), but I also don't think it's a big deal to get all the nodes.

In case you feel strongly that this (fetching the Node directly) approach is better, let me know and I'll adjust the code.

Reply:
nodes are cluster scoped

Reply:
Make sense.
I've saw in our code an example where we use namespace, so I've wrongly assumed it's needed, I guess it's likely ignored in cluster scope.